README.md
# Prometheus MongoDB query exporter
[![.github/workflows/action.yml](https://github.com/raffis/mongodb-query-exporter/workflows/.github/workflows/action.yml/badge.svg)](https://github.com/raffis/mongodb-query-exporter/actions)
[![Go Report Card](https://goreportcard.com/badge/github.com/raffis/mongodb-query-exporter)](https://goreportcard.com/report/github.com/raffis/mongodb-query-exporter)
[![PkgGoDev](https://pkg.go.dev/badge/github.com/raffis/mongodb-query-exporter?tab=subdirectories)](https://pkg.go.dev/github.com/raffis/mongodb-query-exporter?tab=subdirectories)
[![Coverage Status](https://coveralls.io/repos/github/raffis/mongodb-query-exporter/badge.svg?branch=master)](https://coveralls.io/github/raffis/mongodb-query-exporter?branch=master)
[![release](https://img.shields.io/github/release/raffis/mongodb-query-exporter/all.svg)](https://github.com/raffis/mongodb-query-exporter/releases)
[![Artifact HUB](https://img.shields.io/endpoint?url=https://artifacthub.io/badge/repository/prometheus-mongodb-query-exporter)](https://artifacthub.io/packages/search?repo=prometheus-mongodb-query-exporter)

MongoDB aggregation query exporter for [Prometheus](https://prometheus.io).

## Features

* Support for gauge metrics
* Pull and Push (Push is only supported for MongoDB >= 3.6)
* Supports multiple MongoDB servers
* Metric caching support

Note that this is not designed to be a replacement for the [MongoDB exporter](https://github.com/percona/mongodb_exporter) to instrument MongoDB internals. This application exports custom MongoDB metrics in the prometheus format based on the queries (aggregations) you want.

## Installation

Get Prometheus MongoDB aggregation query exporter, either as a [binary](https://github.com/raffis/mongodb-query-exporter/releases/latest) or packaged as a [Docker image](https://hub.docker.com/r/raffis/mongodb-query-exporter).

### Helm Chart
For kubernetes users there is an official helm chart for the MongoDB query exporter.
Please read the installation instructions [here](https://github.com/raffis/mongodb-query-exporter/blob/master/chart/prometheus-mongodb-query-exporter/README.md).

## Usage

```
$ mongodb_query_exporter
```

Use the `-help` flag to get help information.

If you use [MongoDB Authorization](https://docs.mongodb.org/manual/core/authorization/), best practices is to create a dedicated readonly user with access to all databases/collections required:

1. Create a user with '*read*' on your database, like the following (*replace username/password/db!*):

```js
db.getSiblingDB("admin").createUser({
    user: "mongodb_query_exporter",
    pwd: "secret",
    roles: [
        { role: "read", db: "mydb" }
    ]
})
```

2. Set environment variable `MONGODB_URI` before starting the exporter:

```bash
export MDBEXPORTER_MONGODB_URI=mongodb://mongodb_query_exporter:secret@localhost:27017
```

Note: The URI is substituted using env variables `${MY_ENV}`, given that you may also pass credentials from other env variables. See the example bellow.

If you use [x.509 Certificates to Authenticate Clients](https://docs.mongodb.com/manual/tutorial/configure-x509-client-authentication/), pass in username and `authMechanism` via [connection options](https://docs.mongodb.com/manual/reference/connection-string/#connections-connection-options) to the MongoDB uri. Eg:

```
mongodb://CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry@localhost:27017/?authMechanism=MONGODB-X509
```

## Credentials from env variables
You can pass in credentials from env variables.

Given the following URI the exporter will look for the ENV variables called `MY_USERNAME` and `MY_PASSWORD` and automatically use them at the referenced position within the URI.
```bash
export MY_USERNAME=mongodb_query_exporter
export MY_PASSWORD=secret
export MDBEXPORTER_MONGODB_URI=mongodb://${MY_USERNAME}:${MY_PASSWORD}@localhost:27017
```

## Access metrics
The metrics are by default exposed at `/metrics`.

```
curl localhost:9412/metrics
```

## Configuration

The exporter is looking for a configuration in `~/.mongodb_query_exporter/config.yaml` and `/etc/mongodb_query_exporter/config.yaml` or if set the path from the env `MDBEXPORTER_CONFIG`.

You may also use env variables to configure the exporter:

| Env variable             | Description                              | Default |
|--------------------------|------------------------------------------|---------|
| MDBEXPORTER_CONFIG       | Custom path for the configuration        | `~/.mongodb_query_exporter/config.yaml` or `/etc/mongodb_query_exporter/config.yaml` |
| MDBEXPORTER_MONGODB_URI  | The MongoDB connection URI               | `mongodb://localhost:27017`
| MDBEXPORTER_MONGODB_QUERY_TIMEOUT | Timeout until a MongoDB operations gets aborted | `10` |
| MDBEXPORTER_LOG_LEVEL    | Log level                                | `warning` |
| MDBEXPORTER_LOG_ENCODING | Log format                               | `json` |
| MDBEXPORTER_BIND         | Bind address for the HTTP server         | `:9412` |
| MDBEXPORTER_METRICSPATH  | Metrics endpoint                         | `/metrics` |

Note if you have multiple MongoDB servers you can inject an env variable for each instead using `MDBEXPORTER_MONGODB_URI`:

1. `MDBEXPORTER_SERVER_0_MONGODB_URI=mongodb://srv1:27017`
2. `MDBEXPORTER_SERVER_1_MONGODB_URI=mongodb://srv2:27017`
3. ...

### Configure your metrics

Since the v1.0.0 release you should use the config version v3.0 to profit from the latest features.
See the configuration version matrix bellow.

Example:
```yaml
version: 3.0
bind: 0.0.0.0:9412
log:
  encoding: json
  level: info
  development: false
  disableCaller: false
global:
  queryTimeout: 3s
  maxConnection: 3
  defaultCache: 0
servers:
- name: main
  uri: mongodb://localhost:27017
aggregations:
- database: mydb
  collection: objects
  servers: [main] #Can also be empty, if empty the metric will be used for every server defined
  metrics:
  - name: myapp_example_simplevalue_total
    type: gauge #Can also be empty, the default is gauge
    help: 'Simple gauge metric'
    value: total
    overrideEmpty: true # if an empty result set is returned..
    emptyValue: 0       # create a metric with value 0
    labels: []
    constLabels: []
  cache: 0
  mode: pull
  pipeline: |
    [
      {"$count":"total"}
    ]
- database: mydb
  collection: queue
  metrics:
  - name: myapp_example_processes_total
    type: gauge
    help: 'The total number of processes in a job queue'
    value: total
    labels: [type,status]
    constLabels: []
  mode: pull
  pipeline: |
    [
      {"$group": {
        "_id":{"status":"$status","name":"$class"},
        "total":{"$sum":1}
      }},
      {"$project":{
        "_id":0,
        "type":"$_id.name",
        "total":"$total",
        "status": {
          "$switch": {
              "branches": [
                 { "case": { "$eq": ["$_id.status", 0] }, "then": "waiting" },
                 { "case": { "$eq": ["$_id.status", 1] }, "then": "postponed" },
                 { "case": { "$eq": ["$_id.status", 2] }, "then": "processing" },
                 { "case": { "$eq": ["$_id.status", 3] }, "then": "done" },
                 { "case": { "$eq": ["$_id.status", 4] }, "then": "failed" },
                 { "case": { "$eq": ["$_id.status", 5] }, "then": "canceled" },
                 { "case": { "$eq": ["$_id.status", 6] }, "then": "timeout" }
              ],
              "default": "unknown"
          }}
      }}
    ]
- database: mydb
  collection: events
  metrics:
  - name: myapp_events_total
    type: gauge
    help: 'The total number of events (created 1h ago or newer)'
    value: count
    labels: [type]
    constLabels: []
  mode: pull
  # Note $$NOW is only supported in MongoDB >= 4.2
  pipeline: |
    [
      { "$sort": { "created": -1 }},
      {"$limit": 100000},
      {"$match":{
        "$expr": {
          "$gte": [
            "$created",
            {
              "$subtract": ["$$NOW", 3600000]
            }
          ]
        }
      }},
      {"$group": {
        "_id":{"type":"$type"},
        "count":{"$sum":1}
      }},
      {"$project":{
        "_id":0,
        "type":"$_id.type",
        "count":"$count"
      }}
    ]
```

See more examples in the `/example` folder.

## Supported config versions

| Config version           | Supported since   |
|--------------------------|-------------------|
| `v3.0`                   | v1.0.0            |
| `v2.0`                   | v1.0.0-beta5      |
| `v1.0`                   | v1.0.0-beta1      |


## Cache & Push
Prometheus is designed to scrape metrics. During each scrape the mongodb-query-exporter will evaluate all configured metrics.
If you have expensive queries there is an option to cache the aggregation result by setting a cache ttl.
However it is more effective to **avoid cache** and design good aggregation pipelines. In some cases a different scrape interval might also be a solution.
For individual aggregations and/or MongoDB servers older than 3.6 it might still be a good option though.

A better approach is using push instead a static cache, see bellow.

Example:
```yaml
aggregations:
- metrics:
  - name: myapp_example_simplevalue_total
    help: 'Simple gauge metric which is cached for 5min'
    value: total
  servers: [main]
  mode: pull
  cache: 5m
  database: mydb
  collection: objects
  pipeline: |
    [
      {"$count":"total"}
    ]
```

To reduce load on the MongoDB server (and also scrape time) there is a push mode. Push automatically caches the metric at scrape time preferred (If no cache ttl is set). However the cache for a metric with mode push
will be invalidated automatically if anything changes within the configured MongoDB collection. Meaning the aggregation will only be executed if there have been changes during scrape intervals.

>**Note**: This requires at least MongoDB 3.6.

Example:
```yaml
aggregations:
- metrics:
  - name: myapp_example_simplevalue_total
    help: 'Simple gauge metric'
    value: total
  servers: [main]
  # With the mode push the pipeline is only executed if a change occured on the collection called objects
  mode: push
  database: mydb
  collection: objects
  pipeline: |
    [
      {"$count":"total"}
    ]
```

## Debug
The mongodb-query-exporters also publishes a counter metric called `mongodb_query_exporter_query_total` which counts query results for each configured aggregation.
Furthermore you might increase the log level to get more insight.

## Used by
* The balloon helm chart implements the mongodb-query-exporter to expose general stats from the MongoDB like the number of total nodes or files stored internally or externally.
See the [config-map here](https://github.com/gyselroth/balloon-helm/blob/master/unstable/balloon/charts/balloon-mongodb-metrics/templates/config-map.yaml).


Please submit a PR if your project should be listed here!


CHANGELOG.md
# 1.0.0
**Date**: 2022-03-18

Out of beta!
Notably the config format v3.0 was introduced in this release. See readme.md for more information.

## Features
* Support for env substitution within mongodb uri
* Support multiple metric values per aggregation pipeline #24

## Changes
* update: mongodb driver 1.8.4 #38 @fredmaggiowski
* Log a warning if no metrics are configured
* Upgrade to go1.17
* Add matrix mongodb version integration tests
* Various code improvements

## Bugfixes
* Exporter works now even without any servers configured in the config and falls back to the uri env or localhost
* Fallback to default logger if an empty encoder and/or level is configured

## Packaging
* Add support for podMonitor in helm chart
* kustomize base #29
* Add e2e tests and improved pipeline
* Add matrix mongodb version e2e tests
* Publish container image to ghcr.io instead of docker hub


# 1.0.0-beta8
**Date**: 2021-11-04

## Features
* Add an option to override value for empty result set #32 @guillaumelecerf

## Changes
* Helper functions for shared chart labels #33


# 1.0.0-beta7
**Date**: 2021-01-18

## Bugfixes
* Fixes missing Path in ServiceMonitor helm chart

## Features
* Support changeable metrics path #19


# 1.0.0-beta6
**Date**: 2020-11-12

## Bugfixes
* Protect cache from concurrent access

## Changes
* User default prometheus registry instead a custom on


# 1.0.0-beta5
**Date**: 2020-11-11

## Features
* Support for multiple MongoDB servers
* Support for versioned configurations
* Redesigned GO public API
* New log implementation, the default is now json log format
* Added new counter metric for internal stats about quries `mongodb_query_exporter_query_total`
* Added root hint to go to /metrics

## Changes
* Dropped interval pulling mechanism, pull is now synchronous while getting called by an http request
* Drop support for counters (use gauge instead)
* No internal tracking of upstreams anymore, using prometheus ConstMetric
* go 1.15 update

## Packaging
* Provide Helm chart
* Migrated to github actions (from travis-ci)


# 1.0.0-beta4
**Date**: 2020-03-24

## Bugfixes
* Do not abort if lookup config in home fails
* Do not abort pull listeners if an error occurs #2
* Do not panic if MongoDB is unreachable during bootstrap

## Changes
* Execute docker container rootless by default
* introduce two config changes: cacheTime => interval / realtime => mod [pull|push]

## Packaging
* Do not build on travis and in Dockerfile, only build docker image on travis which contains everything else


# 1.0.0-beta3
**Date**: 2020-03-13

## Bugfixes
* Metrics never updated within startPullListeners? #1
* Proper usage of defaultCacheTime from mongodb.defaultCacheTime
* fixes logging output during testing

## Changes
* go 1.13 update


## Packaging
* correctly push build to github releases from travis


cmd/root.go
package cmd

import (
	"fmt"
	"net"
	"net/http"
	"os"
	"os/user"
	"time"

	"github.com/raffis/mongodb-query-exporter/collector"
	"github.com/raffis/mongodb-query-exporter/config"
	v1 "github.com/raffis/mongodb-query-exporter/config/v1"
	v2 "github.com/raffis/mongodb-query-exporter/config/v2"
	v3 "github.com/raffis/mongodb-query-exporter/config/v3"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var (
	configPath    string
	logLevel      string
	logEncoding   string
	bind          string
	uri           string
	metricsPath   string
	queryTimeout  time.Duration
	srv           *http.Server
	promCollector *collector.Collector

	rootCmd = &cobra.Command{
		Use:   "mongodb_query_exporter",
		Short: "MongoDB aggregation exporter for prometheus",
		Long:  `Export aggregations from MongoDB as prometheus metrics.`,
		Run: func(cmd *cobra.Command, args []string) {
			c, conf, err := buildCollector()
			if err != nil {
				panic(err)
			}

			//reg := prometheus.NewRegistry()
			prometheus.MustRegister(c)
			promCollector = c
			c.StartCacheInvalidator()
			srv = buildHTTPServer(prometheus.DefaultGatherer, conf)
			err = srv.ListenAndServe()

			// Only panic if we have a net error
			if _, ok := err.(*net.OpError); ok {
				panic(err)
			} else {
				os.Stderr.WriteString(err.Error() + "\n")
			}
		},
	}
)

func buildCollector() (*collector.Collector, config.Config, error) {
	var configVersion float32
	err := viper.UnmarshalKey("version", &configVersion)
	if err != nil {
		panic(err)
	}

	var conf config.Config
	switch configVersion {
	case 3.0:
		conf = &v3.Config{}

	case 2.0:
		conf = &v2.Config{}

	default:
		conf = &v1.Config{}
	}

	err = viper.Unmarshal(&conf)
	if err != nil {
		panic(err)
	}

	if os.Getenv("MDBEXPORTER_MONGODB_URI") != "" {
		os.Setenv("MDBEXPORTER_SERVER_0_MONGODB_URI", os.Getenv("MDBEXPORTER_MONGODB_URI"))
	}

	if uri != "" && uri != "mongodb://localhost:27017" {
		os.Setenv("MDBEXPORTER_SERVER_0_MONGODB_URI", uri)
	}

	c, err := conf.Build()
	return c, conf, err
}

// Run executes a blocking http server. Starts the http listener with the metrics and healthz endpoints.
func buildHTTPServer(reg prometheus.Gatherer, conf config.Config) *http.Server {
	mux := http.NewServeMux()

	if conf.GetMetricsPath() != "/" {
		mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
			http.Error(w, fmt.Sprintf("Use the %s endpoint", conf.GetMetricsPath()), http.StatusOK)
		})
	}

	mux.HandleFunc(config.HealthzPath, func(w http.ResponseWriter, r *http.Request) { http.Error(w, "OK", http.StatusOK) })
	mux.HandleFunc(conf.GetMetricsPath(), func(w http.ResponseWriter, r *http.Request) {
		promhttp.HandlerFor(reg, promhttp.HandlerOpts{}).ServeHTTP(w, r)
	})

	srv := http.Server{Addr: conf.GetBindAddr(), Handler: mux}
	return &srv
}

// Executes the root command.
func Execute() error {
	return rootCmd.Execute()
}

func init() {
	cobra.OnInitialize(initConfig)
	rootCmd.PersistentFlags().StringVarP(&uri, "uri", "u", config.DefaultMongoDBURI, "MongoDB URI (default is mongodb://localhost:27017). Use MDBEXPORTER_SERVER_%d_MONGODB_URI envs if you target multiple server")
	rootCmd.PersistentFlags().StringVarP(&configPath, "file", "f", "", "config file (default is $HOME/.mongodb_query_exporter/config.yaml)")
	rootCmd.PersistentFlags().StringVarP(&logLevel, "log-level", "l", config.DefaultLogLevel, "Define the log level (default is warning) [debug,info,warn,error]")
	rootCmd.PersistentFlags().StringVarP(&logEncoding, "log-encoding", "e", config.DefaultLogEncoder, "Define the log format (default is json) [json,console]")
	rootCmd.PersistentFlags().StringVarP(&bind, "bind", "b", config.DefaultBindAddr, "Address to bind http server (default is :9412)")
	rootCmd.PersistentFlags().StringVarP(&metricsPath, "path", "p", config.DefaultMetricsPath, "Metric path (default is /metrics)")
	rootCmd.PersistentFlags().DurationVarP(&queryTimeout, "query-timeout", "t", config.DefaultQueryTimeout, "Timeout for MongoDB queries")
	viper.BindPFlag("log.level", rootCmd.PersistentFlags().Lookup("log-level"))
	viper.BindPFlag("log.encoding", rootCmd.PersistentFlags().Lookup("log-encoding"))
	viper.BindPFlag("bind", rootCmd.PersistentFlags().Lookup("bind"))
	viper.BindPFlag("metricsPath", rootCmd.PersistentFlags().Lookup("path"))
	viper.BindPFlag("mongodb.uri", rootCmd.PersistentFlags().Lookup("uri"))
	viper.BindPFlag("mongodb.queryTimeout", rootCmd.PersistentFlags().Lookup("query-timeout"))
	viper.BindEnv("mongodb.uri", "MDBEXPORTER_MONGODB_URI")
	viper.BindEnv("global.queryTimeout", "MDBEXPORTER_MONGODB_QUERY_TIMEOUT")
	viper.BindEnv("log.level", "MDBEXPORTER_LOG_LEVEL")
	viper.BindEnv("log.encoding", "MDBEXPORTER_LOG_ENCODING")
	viper.BindEnv("bind", "MDBEXPORTER_BIND")
	viper.BindEnv("metricsPath", "MDBEXPORTER_METRICSPATH")
}

func initConfig() {
	envPath := os.Getenv("MDBEXPORTER_CONFIG")

	if configPath != "" {
		// Use config file from the flag.
		viper.SetConfigFile(configPath)
	} else if envPath != "" {
		// Use config file from env.
		viper.SetConfigFile(envPath)
	} else {
		// Find home directory.
		usr, err := user.Current()
		if err != nil {
			panic(err)
		}

		// System wide config
		viper.AddConfigPath("/etc/mongodb_query_exporter")
		// Search config in home directory with name ".mongodb_query_exporter" (without extension).
		viper.AddConfigPath(usr.HomeDir + "/.mongodb_query_exporter")
		//config file name without extension
		//	viper.SetConfigName("config")
	}

	if err := viper.ReadInConfig(); err != nil {
		panic(err)
	}
}
